admin/tests/test_integration.py

```diff
@@ -430,3 +432,111 @@ def test_sdconfig_enable_https_on_source_interface():
         data = fobj.read()
     assert HTTPS_OUTPUT == data
+
+
+# The following is the minimal git configuration which can be used to fetch
+# from the SecureDrop Github repository. We want to use this because the
+# developers may have the git setup to fetch from git@github.com: instead
+# of the https, and that requires authentication information.
+GIT_CONFIG = u'''[core]
+        repositoryformatversion = 0
+        filemode = true
+        bare = false
+        logallrefupdates = true
+[remote "origin"]
+        url = https://github.com/freedomofpress/securedrop.git
+        fetch = +refs/heads/*:refs/remotes/origin/*
+'''
+
+ORIGINAL_DIR = os.getcwd()
+NEW_TMP_DIR = ""
+
+
+# This class is to test all the git related operations.
+class TestGitOperations:
+
+    # We will create a new directory and copy the whole git repo
+    # there. This will help us not to destroy the actual work environment.
+    def setup_method(self, method):
+        global NEW_TMP_DIR
+        NEW_TMP_DIR = tempfile.mkdtemp()
+
+        # Clone the SecureDrop repository into a temp directory and move there.
+        cmd = ['git', 'clone',
+               'https://github.com/freedomofpress/securedrop.git']
+        subprocess.check_call(cmd)
+        subprocess.check_call(["mv", "securedrop", NEW_TMP_DIR])
+        os.chdir(os.path.join(NEW_TMP_DIR, 'securedrop/admin'))
+        subprocess.check_call('git reset --hard'.split())
+        # Now we will put in our own git configuration
+        with io.open('../.git/config', 'w') as fobj:
+            fobj.write(GIT_CONFIG)
+        # Let us move to an older tag
+        subprocess.check_call('git checkout 0.6'.split())
+
+    # We will go back to the original working directory when we are done.
+    def teardown_method(self, method):
+        global NEW_TMP_DIR
+        os.chdir(ORIGINAL_DIR)
+        # Let us remove the temporary directory
+        subprocess.check_call('rm -rf {0}'.format(NEW_TMP_DIR).split())
+
+    def test_check_for_update_when_updates_needed(self):
+        cmd = os.path.join(os.path.dirname(CURRENT_DIR),
+                           'securedrop_admin/__init__.py')
+        ansible_base = os.path.join(NEW_TMP_DIR,
+                                    'securedrop/install_files/ansible-base')
+        fullcmd = 'python {0} --root {1} check_for_updates'.format(
+                  cmd, ansible_base)
+        child = pexpect.spawn(fullcmd)
+        child.expect('Update needed', timeout=20)
+
+        child.expect(pexpect.EOF, timeout=10)  # Wait for CLI to exit
+        child.close()
+        assert child.exitstatus == 0
+        assert child.signalstatus is None
+
+    def test_check_for_update_when_updates_not_needed(self):
+        # Determine latest production tag using GitHub release object
+        github_url = 'https://api.github.com/repos/freedomofpress/securedrop/releases/latest'  # noqa: E501
+        latest_release = requests.get(github_url).json()
```

Comment:
A thought, not a request or a blocker by any means. My experience with the GitHub API is that it is unreliable and may fail at random times. Nothing serious but it requires retries. I also do not think it is worth adding the retry in advance. When and if this fails, the retry can be added.